client/network/src/discovery.rs

```diff
@@ -577,37 +578,27 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 				conn,
 				endpoint,
 				failed_addresses,
+				other_established,
 			)
 		}
 	}
 
-	fn inject_connected(&mut self, peer_id: &PeerId) {
-		for k in self.kademlias.values_mut() {
-			NetworkBehaviour::inject_connected(k, peer_id)
-		}
-	}
-
 	fn inject_connection_closed(
 		&mut self,
 		_peer_id: &PeerId,
 		_conn: &ConnectionId,
 		_endpoint: &ConnectedPoint,
-		_handler: <Self::ProtocolsHandler as IntoProtocolsHandler>::Handler,
+		_handler: <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
+		_remaining_established: usize,
 	) {
 		self.num_connections -= 1;
 		// NetworkBehaviour::inject_connection_closed on Kademlia<MemoryStore> does nothing.
```

Comment:
This is no longer true.

https://github.com/libp2p/rust-libp2p/blob/b98d23263a480c77cec8f9ee9d79abc8240f2628/protocols/kad/src/behaviour.rs#L1960-L1975

Reply:
Yes, I noticed it. We need to patch libp2p a little bit: https://github.com/libp2p/rust-libp2p/pull/2572

Reply:
Note that https://github.com/libp2p/rust-libp2p/pull/2572 is released with libp2p v0.44.0 @kpp.